### PR TITLE
Disable labels with the controls they are associated with

### LINKED ISSUE #11809: Some graphical elements are not disabled in setting panels when they should
### Steps to reproduce:

1. Open braille setting panel
2. Uncheck "Show cursor"
3. Press Alt+F

### Actual behavior:

At step 2, the edit field "Cursor blink rate (ms)" and the combo boxes "Cursor shape for focus:" and "Cursor shape for review:" are greyed out but not their associated label. For people with low vision, it is not so clear that these items are disabled.
At step 3, the focus movesto "Message timeout (sec)" field, whereas the acelerator key is alt+T, not alt+F.

### Expected behavior:

At step 2., the labels should also be greyed out.
At step 3, the focus should not move since "Cursor shape for focus:" combo-box (having alt+f accelerator) is disabled.

#### NVDA installed/portable/running from source:
Any
#### NVDA version:
2020.3
#### Windows version:
10
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
2020.2 with same issue
#### If addons are disabled, is your problem still occuring?
Yes
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
No

### Note

I have already a prototype branch to fix this issue. Coming soon.

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:

Fixes #11809

### Summary of the issue:

When some UI controls such as combo-box or edit field are disabled (greyed out), the associated label remains enabled. For low vision people this makes it harder to detect that the option has been disabled. Keeping the label enabled also lets the accelerator key active; thus pressing this acccelerator will make the focus jump to the next focusable control that is not disabled. This is an unexpected behaviour.

### Description of how this pull request fixes the issue:

guiHelper.LabeledControlHelper.control now returns an object whose class inherit both from BaseLabeledControl and the wxCtrlClass class that was passed to the constructor.
This object behaves the same way as the wxCtrlClass except for Enable and Disable methods.
The new BaseLabeledControl class contains the code to manage enabling and disabling the (main) control as well as the associated label. See the example in the BaseLabeledControl doc string for usage.
Thus any labeled controls that was created by guiHelper.LabeledControlHelper in the UI disables or enables its associated label along with itself.

### Testing performed:

In the Braille settings panel:
* Played with the "Show cursor" and "Blink" options and checked visually in the 3 options below that the labels were disabled and enabled along with the control they are associated with.
* For these 3 options, tried their access key and checked that the focus only moves if the option is enabled.

### Notes

I am not an expert in advanced concepts ofOOP Python object programming. Thus, any comment to refine my code is welcome. Especially:
* The check in the __init__ method of BaseLabeledControl may be an indication that I did not use the appropriate OOP tools. Maybe abstract classes? I am not used to them.
* Passing wxCtrlClass to the __init__ method of BaseLabeledControl seems a bit strange to me. Indeed, BaseLabeledControl should always be subclassed to a class that should also inherit from wxCtrlClass. Thus the information seems rendundant. Any other way to do?

Moreover the group name in the advanced setting panel are not greyed out with the rest of the panel. I do not know how to grey it out and, reading the Wx Python documentation, it seems this is an intended behaviour. So I did not dig further.

### Known issues with pull request:

None

### Change log entry:

Not needed.

Cc @feerenrut for review.

### CODE REVIEW

File: source/gui/guiHelper.py [hunk 1]
```diff
@@ -168,9 +214,16 @@ def __init__(self, parent, labelText, wxCtrlClass, **kwargs):
...
+		# Dynamically create a class inheriting from BaseLabeledControl and wxCtrlClass.
+		# This class acts as wxCtrlClass but overloads the Enable and Disable methods
+		# in order to also enable or disable self._label at the same time as the control.
+		LabeledControl = type(
+			f'Labeled{wxCtrlClass.__name__}',
+			(BaseLabeledControl, wxCtrlClass),
+			{},
+		)
+		self._ctrl = LabeledControl(parent, labelText=labelText, wxCtrlClass=wxCtrlClass, **kwargs)
+		self._sizer = associateElements(self._ctrl.label, self._ctrl)
```
Comment (feerrenrut) [inline source/gui/guiHelper.py:227]: Instead you might do something like the following in class `LabeledControlHelper`. Note, I haven't tested the following code. Though it should get you most of the way there. Essentially, at runtime it creates a new subclass of the wx control, with `Enable` and `Disable` overridden. These then trigger a custom event before calling their parent implementations. A listener is bound to this event which will enable / disable the label.

In the past I've run into issues with wx trying to call handlers on destroyed objects, so we unbind if the label is destroyed.

``` python
	import wx.lib.newevent
	# When the control is enabled / disabled this event is raised.
	# A handler is automatically added to the control to ensure the label is also enabled/disabled.
	EnableChanged, EVT_ENABLE_CHANGED = wx.lib.newevent.NewEvent()

	def __init__(self, parent: wx.Window, labelText: str, wxCtrlClass: wx.Control, **kwargs):
		""" @param parent: An instance of the parent wx window. EG wx.Dialog
			@param labelText: The text to associate with a wx control.
			@param wxCtrlClass: The class to associate with the label, eg: wx.TextCtrl
			@param kwargs: The keyword arguments used to instantiate the wxCtrlClass
		"""
		object.__init__(self)
		self._label = _label = wx.StaticText(parent, label=labelText)

		class wxCtrlWithEnableEvnt(wxCtrlClass):
			def Enable(self, enable=True):
				evt = LabeledControlHelper.EnableChanged(isEnabled=enable)
				wx.PostEvent(self, evt)
				super().Enable(enable)

			def Disable(self):
				evt = LabeledControlHelper.EnableChanged(isEnabled=False)
				wx.PostEvent(self, evt)
				super().Disable()

		def _onEnableChanged(evt: wx.Event):
			_label.Enable(evt.isEnabled)

		_ctrl: wx.Control = wxCtrlWithEnableEvnt(parent, **kwargs)
		self._ctrl = _ctrl
		_ctrl.Bind(LabeledControlHelper.EVT_ENABLE_CHANGED, _onEnableChanged)
		# prevent _ctrl trying to call events on destroyed objects
		self._label.Bind(
			wx.wxEVT_DESTROY,
			lambda evt: _ctrl.Unbind(
				LabeledControlHelper.EVT_ENABLE_CHANGED,
				handler=_onEnableChanged
			)
		)
		self._sizer = associateElements(_label, _ctrl)
```

File: source/gui/guiHelper.py [hunk 2]
```diff
@@ -154,6 +154,52 @@ def associateElements( firstElement, secondElement):
...
+		self.label = wx.StaticText(parent, label=labelText)
+		if self.__class__ == BaseLabeledControl:
+			raise TypeError(
+				f"BaseLabeledControl should be subclassed and should also inherit from {wxCtrlClass.__name__}."
+			)
+		# Check that the final class also inherits from wxCtrlClass
+		if not isinstance(self, wxCtrlClass):
+			raise TypeError(f"{self.__class__.__name__} should be a subclass of {wxCtrlClass.__name__}.")
+		self.controlClass = wxCtrlClass
+		self.controlClass.__init__(self, parent, **kwargs)
```
Comment (feerrenrut) [inline source/gui/guiHelper.py:192]: I don't think this is quite right. I think it means that `__new__` won't get called on `wxCtrlClass` which could result in hard to diagnose issues.

**Review by feerrenrut — COMMENTED** [review]
Have you considered using events for this? I think it might be simpler. I had a quick look, but didn't find an event for enable/disable, if there is not you could use a custom event.

See https://docs.wxpython.org/events_overview.html#events-and-event-handling

**Comment by CyrilleB79** [comment]
@feerrenrut thanks for this review and the much better code suggestion. My initial code was clearly not correct regarding the missed __new__ method.

On a functional point of view, I have just added a try/catch in the _onDestroy handler called when the label is destroyed:
When performing tests without this handler, I have encountered a RunTimeError. I have imagined that its cause was that the control was destroyed before the label, and thus _ctrl.Unbind was not working anymore.
* Does this analysis seem correct to you?
* In the except clause, should I put something else than just a `pass` statement? (`log.debug`, `log.debugwarning`)

**Comment by feerrenrut** [comment]
> Does this analysis seem correct to you?

Yes, it seems likely. The destroy event being used like this seems to be making things more complicated. It's a shame there is no 'isDestroyed' method on a `wx.Window`, let's add one to the label and then the handler for `enableChanged` can check if the label has already been destroyed. 

``` python

		class LabelEnableChangedListener(wx.StaticText):
			isDestroyed = False
			isListening = False

			def _onDestroy(self, evt: wx.WindowDestroyEvent):
				self.isDestroyed = True

			def listenForEnableChanged(self, _ctrl: wx.Window):
				self.Bind(wx.EVT_WINDOW_DESTROY, self._onDestroy)
				_ctrl.Bind(LabeledControlHelper.EVT_ENABLE_CHANGED, self._onEnableChanged)
				self.isListening = True

			def _onEnableChanged(self, evt: wx.Event):
				if self.isListening and not self.isDestroyed:
					self.Enable(evt.isEnabled)

		class WxCtrlWithEnableEvnt(wxCtrlClass):
			def Enable(self, enable=True):
				evt = LabeledControlHelper.EnableChanged(isEnabled=enable)
				wx.PostEvent(self, evt)
				super().Enable(enable)

			def Disable(self):
				evt = LabeledControlHelper.EnableChanged(isEnabled=False)
				wx.PostEvent(self, evt)
				super().Disable()

		self._label = LabelEnableChangedListener(parent, label=labelText)
		self._ctrl = WxCtrlWithEnableEvnt(parent, **kwargs)
		self._label.listenForEnableChanged(self._ctrl)
		self._sizer = associateElements(self._label, self._ctrl)
```

**Comment by CyrilleB79** [comment]
Not sure to understand.
Is there still an issue in my last code proposal (commit f3558d5)? The try/catch was aimed to avoid _ctrl.Unbind failing.

Are you fixing something else in your new proposal (in https://github.com/nvaccess/nvda/pull/11812#issuecomment-731982881)?

Or is it just to organize the code better, defining a subclass of wx.Label rather than binding sparse functions with events?

No problem to modify my code following your last suggestion. However, I just want to be sure to fully understand what I am committing!

**Comment by feerrenrut** [comment]
Yes, this is just an alternative approach. I think it's more tidy, it means that the order of destruction doesn't matter anymore.

The core of the problem is understanding that the lifetime of the wx (c++ object) does not necessarily match the lifetime of the python object. I figure this is a neater solution because it doesn't tie the two python objects together in any additional way, like my first suggestion with the introduction of the label listening to the destroy event of the control. This new approach ensure that the label is only watching the control, but will not attempt to modify itself if the C++ object has been destroyed already.

**Comment by CyrilleB79** [comment]
Thanks for this clear and detailed explanation.
The code has been updated according to this last suggestion.

**Review by feerrenrut — APPROVED** [review]
Thanks @CyrilleB79

**Pull request merged**